src/assign.c

```diff
@@ -1133,8 +1133,14 @@ void writeNA(SEXP v, const int from, const int n)
     // If there's ever a way added to R API to pass NA_STRING to allocVector() to tell it to initialize with NA not "", would be great
     for (int i=from; i<=to; ++i) SET_STRING_ELT(v, i, NA_STRING);
     break;
-  case VECSXP: case EXPRSXP :
-    // although allocVector already initializes to R_NilValue, we use writeNA() in other places too, so we shouldn't skip this assign
+  case VECSXP:
+    for (int i=from; i<=to; ++i) {
+      SEXP na_scalar = allocVector(LGLSXP, 1);
+      LOGICAL(na_scalar)[0] = NA_LOGICAL;
+      SET_VECTOR_ELT(v, i, na_scalar);
+    }
```

Comment:
I was going to suggest, if we go ahead with this change, using `R_LogicalNAValue` here. That's R's internal constant object to save the multiple allocations (R has `R_TrueValue` and `R_FalseValue` too). However, then I remembered packages don't have access to those internals.  We can use `ScalarLogical(NA_LOGICAL)`, though, and that returns `R_LogicalNAValue` without allocating. So this loop would become :
```
SEXP na = ScalarLogical(NA_LOGICAL);
for (int i=from; i<=to; ++i) SET_VECTOR_ELT(v, i, na);
```
However, lets say `:=` or `set` was then used to write FALSE into some of those list column cells containing NA. That would have the potential to change R's global constant (that has happened before).
However2, we're talking about items within a list column cell here, and I suspect `:=` and `set` don't have the ability to update that nested level. What happened before was with 1-row data.table's containing a logical column: that one-row logical column could be `R_[True|False|LogicalNA]Value` and a `:=` on such a one-row column did corrupt R's internal value. So we catch that now by reallocating 1-row logical columns.
However3, if `:=` and `set` don't already detect and prevent writing to `R_[True|False|LogicalNA]Value`, we should put that in regardless (they could then either change the pointer rather than the contents of the length-1 logical, or they could allocate a new length-1 logical if an attribute is being attached by user.)
So, in summary, after writing out loud here, I would rule out this loop as it stands now for the reason of allocating all those length-1 objects could gobble memory in large cases. Which is maybe why we used empty list (R_NilValue).  If we go ahead with the NULL to NA change, we could review `set()` and `:=` w.r.t. length-1 logicals, and if ok, use `ScalarLogical(NA_LOGICAL)` here.
Maybe even the recent `att="t"` attached to `R_FalseValue` problem could be resolved in a different more robust way inside `set`/`:=` rather than what I did.

Reply:
I was thinking about the potential overhead in allocating a lot of length-1 logical vectors as well. I agree that using R_LogicalNA seems like a great / more efficient alternative. I did not know that `:=` and `set` could change that, but yes I agree that seems like something `:=` / `set` should detect/handle, and that should not stop us from using a more efficient code here.